kotlinx-coroutines-test/common/src/internal/TestMainDispatcher.kt

```diff
@@ -61,29 +61,31 @@ internal class TestMainDispatcher(delegate: CoroutineDispatcher):
      * next modification.
      */
     private class NonConcurrentlyModifiable<T>(initialValue: T, private val name: String) {
-        private val readers = atomic(0) // number of concurrent readers
-        private val isWriting = atomic(false) // a modification is happening currently
+        private val reader: AtomicRef<Throwable?> = atomic(null) // reader currently accessing value
+        private val writer: AtomicRef<Throwable?> = atomic(null) // writer currently performing value modification
         private val exceptionWhenReading: AtomicRef<Throwable?> = atomic(null) // exception from reading
         private val _value = atomic(initialValue) // the backing field for the value
 
-        private fun concurrentWW() = IllegalStateException("$name is modified concurrently")
-        private fun concurrentRW() = IllegalStateException("$name is used concurrently with setting it")
+        private fun concurrentWW(location: Throwable) = IllegalStateException("$name is modified concurrently", location)
+        private fun concurrentRW(location: Throwable) = IllegalStateException("$name is used concurrently with setting it", location)
 
         var value: T
             get() {
-                readers.incrementAndGet()
-                if (isWriting.value) exceptionWhenReading.value = concurrentRW()
+                val readerLocation = Throwable()
+                reader.value = readerLocation
+                writer.value?.let { exceptionWhenReading.value = concurrentRW(it) }
                 val result = _value.value
-                readers.decrementAndGet()
+                reader.compareAndSet(readerLocation, null)
```

Comment:
Imagine two reads and one write happening in parallel. Clearly, something we should report. However, the following may happen:
* The first read sets its location.
* The second read sets its location, overwriting the first.
* The second read completes, setting the `reader` to null.
* The write starts and finishes, despite the first read still not having finished.

We want to catch as many errors as possible. Here's an idea: what if we keep the counter, but also, before incrementing the counter, we write a `Throwable` corresponding to the reader's location to a separate field called, let's say, `lastReadLocation`, which never becomes `null` again? If looking at the counter, the writer detects that there are concurrent reads, it will go to that field and take out the throwable.

Reply:
Yeah, the impl was already leaky.  I figured it was attempting to do a best-effort approach to trapping these conditions while avoiding significantly changing threading behavior through locks.  I like your idea, however.  I'll try to get some free cycles to update the PR to reflect this.

Reply:
Ok, I've updated the PR to retain the reader counter but add a last-reader reference for reporting purposes.  Does this look like what you were after?